core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -815,73 +841,81 @@ public void listStatus(AlluxioURI path, ListStatusContext context,
+          auditContext.setAllowed(false);
+          throw e;
+        }
 
-          boolean isLoaded = inodeDirectory.isDirectChildrenLoaded();
-          if (context.getOptions().getRecursive()) {
-            isLoaded = areDescendantsLoaded(inodeDirectory);
-          }
-          if (isLoaded) {
-            // no need to load again.
-            loadMetadataContext.getOptions().setLoadDescendantType(LoadDescendantPType.NONE);
+        DescendantType loadDescendantType;
+        if (context.getOptions().getLoadMetadataType() == LoadMetadataPType.NEVER) {
+          loadDescendantType = DescendantType.NONE;
+        } else if (context.getOptions().getRecursive()) {
+          loadDescendantType = DescendantType.ALL;
+        } else {
+          loadDescendantType = DescendantType.ONE;
+        }
+        // load metadata for 1 level of descendants, or all descendants if recursive
+        LoadMetadataContext loadMetadataContext = LoadMetadataContext.mergeFrom(
+            LoadMetadataPOptions.newBuilder()
+                .setCreateAncestors(true)
+                .setLoadDescendantType(GrpcUtils.toProto(loadDescendantType))
+                .setCommonOptions(FileSystemMasterCommonPOptions.newBuilder()
+                    .setTtl(context.getOptions().getCommonOptions().getTtl())
+                    .setTtlAction(context.getOptions().getCommonOptions().getTtlAction())));
+        Inode inode;
+        if (inodePath.fullPathExists()) {
+          inode = inodePath.getInode();
+          if (inode.isDirectory()
+              && context.getOptions().getLoadMetadataType() != LoadMetadataPType.ALWAYS) {
+            InodeDirectory inodeDirectory = inode.asDirectory();
+
+            boolean isLoaded = inodeDirectory.isDirectChildrenLoaded();
+            if (context.getOptions().getRecursive()) {
+              isLoaded = areDescendantsLoaded(inodeDirectory);
+            }
+            if (isLoaded) {
+              // no need to load again.
+              loadMetadataContext.getOptions().setLoadDescendantType(LoadDescendantPType.NONE);
+            }
           }
+        } else {
+          checkLoadMetadataOptions(context.getOptions().getLoadMetadataType(), inodePath.getUri());
         }
-      } else {
-        checkLoadMetadataOptions(context.getOptions().getLoadMetadataType(), inodePath.getUri());
-      }
 
-      loadMetadataIfNotExist(rpcContext, inodePath, loadMetadataContext);
-      ensureFullPathAndUpdateCache(inodePath);
-      inode = inodePath.getInode();
-      auditContext.setSrcInode(inode);
-      DescendantType descendantTypeForListStatus =
-          (context.getOptions().getRecursive()) ? DescendantType.ALL : DescendantType.ONE;
-      listStatusInternal(context, rpcContext, inodePath, auditContext, descendantTypeForListStatus,
-          resultStream, 0);
-      auditContext.setSucceeded(true);
-      Metrics.FILE_INFOS_GOT.inc();
+        loadMetadataIfNotExist(rpcContext, inodePath, loadMetadataContext, false);
```

Comment:
Can the new `syncMetadata` also take care of `loadMetadata`? I think the full sync is a superset of load metadata, so can't we just do the `sync`, which can optionally take care of loading metadata only if requested? I think load metadata only happens on getFileInfo and listStatus.

Reply:
Unfortunately in my testing this breaks behavior and some assumptions about what metadata gets loaded during these RPCs. I did this because my goal around these improvements was to not change any assumptions about how we load metadata.

I don't remember exactly which tests broke, but basically loading is a strict load of metadata, whereas a sync could possibly delete or remove inodes, and so that broke a bunch of test assumptions. Breaking previous RPC behavior probably requires a larger discussion.

Reply:
Yeah, I know load is different from sync. But maybe something like `load-only` could be an option to `sync`, so that when that option is used, it does not do the update/delete part of the sync?

This would only be for greater simplicity. I'm not sure if this would actually make it simpler, but I thought maybe it could.

Reply:
That's actually what it already does....go to the definition of loadMetadataIfNotExist :)

Reply:
Yeah, I know `loadMetadata` call reuses sync. However, do you think it is possible for everything to be taken care of with the `syncMetadata()` at the beginning of this method? Basically, for `listStatus` and `fileInfo` they both have a concept of loading metadata, which is just the creating part of sync. I wonder if everything could be taken care of in the `syncMetadata()` call, so we don't even need the `loadMetadata()` method anymore.

(This would not be part of this PR)